llvm/tools/llvm-foreach/llvm-foreach.cpp

```diff
@@ -221,17 +270,18 @@ int main(int argc, char **argv) {
       Args[OutIncrementArg.ArgNum] = IncOutArg;
     }
 
-    std::string ErrMsg;
-    // TODO: Add possibility to execute commands in parallel.
-    int Result =
-        sys::ExecuteAndWait(Prog, Args, /*Env=*/None, /*Redirects=*/None,
-                            /*SecondsToWait=*/0, /*MemoryLimit=*/0, &ErrMsg);
-    if (Result != 0) {
-      errs() << "llvm-foreach: " << ErrMsg << '\n';
-      Res = Result;
-    }
+    // Do not start execution of a new job until previous one(s) are finished,
+    // if the maximum number of parallel workers is reached.
+    while (JobsSubmitted.size() == JobsInParallel)
+      Res = checkIfJobsAreFinished(JobsSubmitted, /*BlockingWait*/ false);
+
+    JobsSubmitted.emplace_back(sys::ExecuteNoWait(
+        Prog, Args, /*Env=*/None, /*Redirects=*/None, /*MemoryLimit=*/0));
```

Comment:
If `Res` above is not zero, new jobs should not probably be started. Instead, all running could be killed and the tool would just exit with non-zero.

Reply:
Yeah, it's a good thought, but it was done on purpose in c69a3115cfef390edabf60a1a7b0ae3321dafd19:

> "we also want to allow for the compilation to continue so the user can use the generated binary knowing the timing issue as stated.").

I believe I have a bug here (should not assign the result to `Res` if it was successful), I'll apply the fix.

Reply:
> "we also want to allow for the compilation to continue so the user can use the generated binary knowing the timing issue as stated."

checkIfJobsAreFinished may return non-zero for a variety of reasons, not only due to a timeout. In fact timeout is not even possible, AFAICS, as SecondsToWait for sys::Wait is 0. So I believe the most likely reason for checkIfJobsAreFinished non-zero return is some error during tool execution, in which case the results will be discarded, and it does not seem to make sense to run any other tools after that.

Reply:
I suggest that we discuss that matter separately and apply some changes (if needed) through a separate PR: apparently there was a decision to **not to stop** submitting further jobs if one of them failed when we executed them in serial manner. It is not clear to me why we shouldn't follow the same logic for parallel execution.

Submitting further jobs after the first failed also allows us to detect *all* issues in *all* submitted jobs and not just in the first one.

Reply:
> Submitting further jobs after the first failed also allows us to detect all issues in all submitted jobs and not just in the first one.

that does not seem right to me, and can be resource-costly approach. All jobs in a llvm-foreach batch are similar and in most cases they will fail for the same reason, I believe.